ulog.cpp

```diff
@@ -0,0 +1,280 @@
+    }
+
+#define NO_FIRST_MSG_OFFSET 255
+
+struct _packed_ ulog_msg_header {
+    uint16_t msg_size;
+    uint8_t msg_type;
+};
+
+static bool _ulog_logging_start_timeout(void *data)
+{
+    ULog *ulog = static_cast<ULog *>(data);
+    return ulog->logging_start_timeout();
+}
+
+bool ULog::logging_start_timeout()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_START;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = TARGET_SYSTEM_ID;
+
+    mavlink_msg_command_long_encode(SYSTEM_ID, 0, &msg, &cmd);
+    _endpoint->write_msg(&msg);
+
+    return true;
+}
+
+static bool _on_master_msg_cb(struct buffer *buf, void *data)
+{
+    ULog *ulog = static_cast<ULog *>(data);
+    return ulog->handle_fligh_stack_message(buf);
+}
+
+bool ULog::start(Mainloop *mainloop, Endpoint *master, const char *base_path)
+{
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    char path[1024];
+
+    null_check(mainloop, false);
+    null_check(master, false);
+
+    if (_file) {
+        log_error("ULog already started");
+        return false;
+    }
+
+    snprintf(path, sizeof(path), "%s/%i-%i-%i_%i-%i-%i.ulg", (base_path ? base_path : ""),
+             timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday, timeinfo->tm_hour,
+             timeinfo->tm_min, timeinfo->tm_sec);
+    _file = fopen(path, "wb");
+    if (!_file) {
+        log_error_errno(errno, "Unable to open ULog file(%s): (%m)", path);
+        return false;
+    }
+
+    _endpoint = master;
+    _mainloop = mainloop;
+    _mainloop->set_on_master_msg_callback(_on_master_msg_cb, this);
+    _timeout = _mainloop->timeout_add(MSEC_PER_SEC, _ulog_logging_start_timeout, this);
+    _waiting_header = true;
+    _waiting_first_msg_offset = false;
+    logging_start_timeout();
+
+    return _timeout;
+}
+
+void ULog::stop()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    if (!_file) {
+        log_error("ULog not started");
+        return;
+    }
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_STOP;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = TARGET_SYSTEM_ID;
+
+    mavlink_msg_command_long_encode(SYSTEM_ID, 0, &msg, &cmd);
+    _endpoint->write_msg(&msg);
+
+    fflush(_file);
+    fclose(_file);
+    _file = NULL;
+    _expected_seq = 0;
+    _mainloop->set_on_master_msg_callback(NULL, NULL);
+}
+
+bool ULog::handle_fligh_stack_message(struct buffer *buffer)
+{
+    const bool mavlink2 = buffer->data[0] == MAVLINK_STX;
+    uint32_t msg_id;
+
+    if (mavlink2) {
+        struct mavlink_router_mavlink2_header *msg
+            = (struct mavlink_router_mavlink2_header *)buffer->data;
+        msg_id = msg->msgid;
+    } else {
+        struct mavlink_router_mavlink1_header *msg
+            = (struct mavlink_router_mavlink1_header *)buffer->data;
+        msg_id = msg->msgid;
```

Comment:
We should output a warning here, and potentially abort since it cannot be used with mavlink 1